asserts/ifacedecls.go

```diff
@@ -0,0 +1,477 @@
+	}
+
+}
+
+// PlugRule holds the rule of what is allowed, wrt installation and connection, for a plug of a specific interface for a snap.
+type PlugRule struct {
+	Interface string
+
+	AllowInstallation *PlugInstallationConstraints
+	DenyInstallation  *PlugInstallationConstraints
+
+	AllowConnection *PlugConnectionConstraints
+	DenyConnection  *PlugConnectionConstraints
+
+	AllowAutoConnection *PlugConnectionConstraints
+	DenyAutoConnection  *PlugConnectionConstraints
+}
+
+func (r *PlugRule) setPlugInstallationConstraints(field string, cstrs *PlugInstallationConstraints) {
+	switch field {
+	case "allow-installation":
+		r.AllowInstallation = cstrs
+	case "deny-installation":
+		r.DenyInstallation = cstrs
+	default:
+		panic("unknown PlugRule field " + field)
+	}
+}
+
+func (r *PlugRule) setPlugConnectionConstraints(field string, cstrs *PlugConnectionConstraints) {
+	switch field {
+	case "allow-connection":
+		r.AllowConnection = cstrs
+	case "deny-connection":
+		r.DenyConnection = cstrs
+	case "allow-auto-connection":
+		r.AllowAutoConnection = cstrs
+	case "deny-auto-connection":
+		r.DenyAutoConnection = cstrs
+	default:
+		panic("unknown PlugRule field " + field)
+	}
+}
+
+// PlugInstallationConstraints specifies a set of constraints on an interface plug relevant to the installation of snap.
+type PlugInstallationConstraints struct {
+	PlugAttributes *AttributeConstraints
+}
+
+func compilePlugInstallationConstraints(interfaceName string, entry string, constraints interface{}) (*PlugInstallationConstraints, error) {
+	context := fmt.Sprintf("%s in plug rule for interface %q", entry, interfaceName)
+	cMap, shortcut, err := checkMapOrShortcut(context, constraints)
+	if err != nil {
+		return nil, err
+	}
+	if cMap == nil {
+		if shortcut {
+			return &PlugInstallationConstraints{PlugAttributes: AlwaysMatchAttributes}, nil
+		}
+		return &PlugInstallationConstraints{PlugAttributes: NeverMatchAttributes}, nil
+	}
+	attrs, err := compileAttributeContraints(cMap["plug-attributes"])
+	if err != nil {
+		return nil, fmt.Errorf("cannot compile plug-attributes in %s: %v", context, err)
+	}
+	return &PlugInstallationConstraints{PlugAttributes: attrs}, nil
+}
+
+// PlugConnectionConstraints specfies a set of constraints on an interface plug for a snap relevant to its connection or auto-connection.
+type PlugConnectionConstraints struct {
+	SlotSnapIDs      []string
+	SlotPublisherIDs []string
+
+	PlugAttributes *AttributeConstraints
+	SlotAttributes *AttributeConstraints
+}
+
+func (c *PlugConnectionConstraints) setAttributeConstraints(field string, cstrs *AttributeConstraints) {
+	switch field {
+	case "plug-attributes":
+		c.PlugAttributes = cstrs
+	case "slot-attributes":
+		c.SlotAttributes = cstrs
+	default:
+		panic("unknown PlugConnectionConstraints field " + field)
+	}
+}
+
+func (c *PlugConnectionConstraints) setIDConstraints(field string, cstrs []string) {
+	switch field {
+	case "slot-snap-id":
```

Comment:
Can we please have slot-snap-type here and plug-snap-type on the other end? Also lists.